prometheus/vec_test.go

```diff
@@ -125,6 +125,86 @@ func testDeleteLabelValues(t *testing.T, vec *GaugeVec) {
 	}
 }
 
+func TestDeletePartialMatch(t *testing.T) {
+	vec := NewGaugeVec(
+		GaugeOpts{
+			Name: "test",
+			Help: "helpless",
+		},
+		[]string{"l1", "l2"},
+	)
+	testDeletePartialMatch(t, vec)
+}
+
+func testDeletePartialMatch(t *testing.T, vec *GaugeVec) {
+	// No metric value is set.
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "v2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c1 := vec.MustCurryWith(Labels{"l1": "v1"})
+	c1.WithLabelValues("2").Inc()
+
+	// Try to delete nonexistent label lx with existent value v1.
+	if got, want := c1.DeletePartialMatch(Labels{"lx": "v1"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Delete with valid pair l1: v1.
+	if got, want := c1.DeletePartialMatch(Labels{"l1": "v1"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with partially invalid labels.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "xv2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with a single valid label which matches multiple metrics.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	vec.With(Labels{"l1": "v1", "l2": "vv22"}).(Gauge).Set(84)
+	c3 := vec.MustCurryWith(Labels{"l2": "l2C3CurriedValue"}) // Used below
+	vec.With(Labels{"l1": "v3", "l2": "v3"}).(Gauge).Set(168)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1"}), 2; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete a value which shouldn't be in our base vector (only the curried one c3).
+	if got, want := vec.DeletePartialMatch(Labels{"l2": "l2C3CurriedValue"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c2 := vec.MustCurryWith(Labels{"l2": "l2CurriedValue"})
+	c2.With(Labels{"l1": "11"}).Inc()
+
+	// Delete with valid curried pair l2: l2CurriedValue.
+	if got, want := c2.DeletePartialMatch(Labels{"l2": "l2CurriedValue"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+	// Try to delete with a label value from before currying.
+	if got, want := c2.DeletePartialMatch(Labels{"l2": "v3"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
```

Comment:
I have a difficult time wrapping my head around the currying behavior, so I added some more tests and would like some clarification here. When we curry, we set the value of a label at some index to some value. Is that now _always_ the definitive value for that index, even if the vector was created before currying? 

In these tests, either line 182 or line 186 can succeed, but not both. In this test flow, we create a metric like
`{"l1": "v3", "l2": "v3"} 168`
Then, we create a new vector c2 where we curry with `{"l2": "l2CurriedValue"}`:
`{"l1": "11", "l2": "l2CurriedValue"} 1`

When it comes time to delete based on the `l2` label, the underlying c2 metrics _still_ contain 
`{"l1": "v3", "l2": "v3"}` and
`{"l1": "11", "l2": "l2CurriedValue"}`

And `curry` contains the `l2CurriedValue` at index 1.

I _think_ I should pretend as if the `v3` value does not exist anymore. Is that correct?

I guess for other deletions this is handled (or at least hidden) by the hash function, but how should it be handled in this case?

Reply:
My memory about the currying is quite distant (I implemented it long ago…), but as far as I do remember, the currying is like a "view" on the underlying original vector where the curried labels behave more or less like constant labels.

From that perspective, I think both line 182 and 186 should be considered invalid as they refer to a label that has been curried away. Note that the `Delete` method errors out if you include the curried label because the cardinality doesn't match anymore. I guess we want to avoid giving `DeletePartialMatch` an error return. In that case, I suggest to treat the curried labels the same as non-existing labels when `DeletePartialMatch` is called on a curried vector.

Reply:
I was thinking it makes sense to include curried labels in `DeletePartialMatch`, otherwise those metrics would still need to be entirely reconstructed in order to delete. Then it also doesn't change the behavior of `DeletePartialMatch` based on the workflow before it is called.

For example, a user could construct two vectors which are identical from the scrape perspective, but one which is curried and another which is not. It would be confusing if they then call `DeletePartialMatch` and only the non-curried vector disappears despite the curried one reporting the same label. With `DeletePartialMatch` the cardinality effectively doesn't matter anymore since users are expected to specify only a subset of the labels to begin with.

But if I'm understanding your response correctly, in any case the `v3` value should be concealed by currying, and calls to delete based on that value should not have any effect.

> From that perspective, I think both line 182 and 186 should be considered invalid as they refer to a label that has been curried away.

Based on the above, I think line 182 would then be the expected behavior (one metric would be deleted because the current representation would be `{"l1": "11", "l2": "l2CurriedValue"}` and the user has asked for deletion of all metrics with `{"l2": "l2CurriedValue"}`)

Reply:
> For example, a user could construct two vectors which are identical from the scrape perspective, but one which is curried and another which is not. It would be confusing if they then call DeletePartialMatch and only the non-curried vector disappears despite the curried one reporting the same label.

I think the terminology is a bit off here. A vector is the whole thing. What you delete are elements in the vector, AKA metrics.

If you curry a vector, the curried vector still refers to the same vector as the uncurried vector. If you delete a metric from the uncurried vector, it will also appear from the curried vector, because they are still the same vector. The curried vector is just a "view" on the uncurried vector.

The doc comment says: "CurryWith returns a vector curried with the provided labels, i.e. the returned vector has those labels pre-set for all labeled operations performed on it." I wouldn't like to change this semantics for the new `DeletePartialMatch`. Therefore, `curriedVec.DeletePartialMatch(Labels{"l1": "11"})` should have the exact same effect as `baseVec.DeletePartialMatch(Labels{"l1": "11", "l2": "l2CurriedValue"})`. This pattern should also be the one you usually use. It could be argued what should happen if the curried label is included. `curriedVec.DeletePartialMatch(Labels{"l1": "11", "l2": "l2CurriedValue"}})` could be seen as redundant and therefore should still have the same effect, while `curriedVec.DeletePartialMatch(Labels{"l1": "11", "l2": "somethingelse"}})` could then be seen as a contradiction, which should be a no-op or even an error. _However_ I think this view is at odds with the behavior of the plain `Delete` where `curriedVec.Delete(Labels{"l1": "11", "l2": "l2CurriedValue"}})` is _not_ seen as redundant but results in an error. Therefore, I prefer to treat a specification of the curried label the same as specifying a non-existing label, resulting in a no-op. (It would be even more consistent to return an error, but I do understand that we wouldn't like an error return just for this one purpose.)


> With DeletePartialMatch the cardinality effectively doesn't matter anymore since users are expected to specify only a subset of the labels to begin with.

Yes, but my point isn't so much that the plain `Delete` returns a cardinality error. It's that the curried label is added implicitly to all labeled operation before applying them to the underlying vector. If we allow a redundant specification of the curried label in `DeletePartialMatch`, we should allow the same for `Delete`, which would also resolve the cardinality error. But that would be a change in semantics (and arguably a confusing one).

Reply:
I've had some other priorities come up but I'm still planning to come back here.

I'll take more time to review your comment here and try to reflect it with test cases. I think I understand where you're coming from now, I guess what's confusing to me is the difference between the user's intent and the implementation when calling `DeletePartialMatch`. If my vector is curried with some value, and I call `DeletePartialMatch` with that label value, I'd expect that I would no longer see metrics containing that value. It might be redundant because by definition all metrics curried with that value will match, but I'm considering cases where a client calls `DeletePartialMatch` on a list of vectors, or where they can't reconstruct the other (uncurried) values for their vector. The alternative would be to call `DeletePartialMatch` on the base vector instead, but that may not be available to a caller.

However, I'm open to implementing it the way you've described. In that case, if only uncurried labels are supported for partial deletion, it may be worth adding an error to remain consistent with `Delete` and differentiate between situations where a user passed a valid label which matched zero metrics and those where a user passed an invalid (perhaps curried) value. What would your preference be on the error?